crates/floresta-wire/src/p2p_wire/tests/chain_selector.rs

```diff
@@ -0,0 +1,537 @@
+
+                match req {
+                    NodeRequest::GetHeaders(hashes) => {
+                        let pos = hashes.first().unwrap();
+                        let pos = self.headers.iter().position(|h| h.block_hash() == *pos);
+                        let headers = match pos {
+                            None => vec![],
+                            Some(pos) => self.headers[(pos + 1)..].to_vec(),
+                        };
+
+                        self.node_tx
+                            .send(NodeNotification::FromPeer(
+                                self.peer_id,
+                                PeerMessages::Headers(headers),
+                            ))
+                            .await
+                            .unwrap();
+                    }
+                    NodeRequest::GetUtreexoState((hash, _)) => {
+                        let filters = self.filters.get(&hash).unwrap().clone();
+                        self.node_tx
+                            .send(NodeNotification::FromPeer(
+                                self.peer_id,
+                                PeerMessages::UtreexoState(filters),
+                            ))
+                            .await
+                            .unwrap();
+                    }
+                    NodeRequest::Shutdown => {
+                        return;
+                    }
+                    NodeRequest::GetBlock((hashes, _)) => {
+                        for hash in hashes {
+                            let block = self.blocks.get(&hash).unwrap().clone();
+                            self.node_tx
+                                .send(NodeNotification::FromPeer(
+                                    self.peer_id,
+                                    PeerMessages::Block(block),
+                                ))
+                                .await
+                                .unwrap();
+                        }
+                    }
+
+                    _ => {}
+                }
+            }
+        }
+    }
+
+    fn create_peer(
+        headers: Vec<Header>,
+        blocks: HashMap<BlockHash, UtreexoBlock>,
+        filters: HashMap<BlockHash, Vec<u8>>,
+        node_sender: Sender<NodeNotification>,
+        sender: Sender<NodeRequest>,
+        node_rcv: Receiver<NodeRequest>,
+        peer_id: u32,
+    ) -> LocalPeerView {
+        let peer = TestPeer::new(node_sender, headers, blocks, filters, node_rcv, peer_id);
+        task::spawn(peer.run());
+
+        LocalPeerView {
+            address: "127.0.0.1".parse().unwrap(),
+            services: ServiceFlags::from(1 << 25),
+            user_agent: "/utreexo:0.1.0/".to_string(),
+            height: 0,
+            state: PeerStatus::Ready,
+            channel: sender,
+            port: 8333,
+            feeler: false,
+            banscore: 0,
+            address_id: 0,
+            _last_message: Instant::now(),
+        }
+    }
+
+    pub async fn setup_test(
```

Comment:
If you do this, you can move the asserts to each function. The way it works currently we can't reuse this function easily.

```diff
     pub async fn setup_test(
-        test_name: &str,
         peers: Vec<(
             Vec<Header>,
             HashMap<BlockHash, UtreexoBlock>,
@@ -282,7 +278,7 @@ mod tests_utils {
         )>,
         pow_fraud_proofs: bool,
         network: floresta_chain::Network,
-    ) {
+    ) -> Arc<ChainState<KvChainStore<'static>>> {
         let datadir = format!("./data/{}.node_test", rand::random::<u32>());
         let chainstore = KvChainStore::new(datadir.clone()).unwrap();
         let mempool = Arc::new(RwLock::new(Mempool::new()));
@@ -348,60 +344,7 @@ mod tests_utils {
             .unwrap()
             .unwrap();
 
-        let headers = get_test_headers();
-
-        match test_name {
-            "test_chain_selector" => {
-                assert_eq!(chain.is_in_idb(), false);
-                assert_eq!(chain.get_best_block().unwrap().0, 2015);
-                assert_eq!(
-                    chain.get_best_block().unwrap().1,
-                    headers[2015].block_hash()
-                );
-            }
-
-            "two_peers_different_tips" => {
-                assert_eq!(chain.is_in_idb(), false);
-                assert_eq!(chain.get_best_block().unwrap().0, 2014);
-                assert_eq!(
-                    chain.get_best_block().unwrap().1,
-                    headers[2014].block_hash()
-                );
-            }
-            "ten_peers_different_tips" => {
-                assert_eq!(chain.is_in_idb(), false);
-                assert_eq!(chain.get_best_block().unwrap().0, 2013);
-                assert_eq!(
-                    chain.get_best_block().unwrap().1,
-                    headers[2013].block_hash()
-                );
-            }
-            "two_peers_one_lying" => {
-                assert_eq!(chain.is_in_idb(), false);
-                assert_eq!(
-                    chain.get_root_hashes()[3],
-                    NodeHash::from_str(
-                        "bfe030a7a994b921fb2329ff085bd0f2351cb5fa251985d6646aaf57954b782b"
-                    )
-                    .unwrap()
-                );
-                assert_eq!(chain.get_root_hashes().len(), 6);
-                assert_eq!(chain.get_best_block().unwrap().1, headers[119].block_hash());
-            }
-            "ten_peers_one_honest" => {
-                assert_eq!(chain.is_in_idb(), false);
-                assert_eq!(
-                    chain.get_root_hashes()[3],
-                    NodeHash::from_str(
-                        "bfe030a7a994b921fb2329ff085bd0f2351cb5fa251985d6646aaf57954b782b"
-                    )
-                    .unwrap()
-                );
-                assert_eq!(chain.get_root_hashes().len(), 6);
-                assert_eq!(chain.get_best_block().unwrap().1, headers[119].block_hash());
-            }
-            _ => {}
-        }
+        chain
     }
```

Reply:
Done!